Guides for Amazon Location Service (#2777)
aws-amplify/docs: docs/guides/location-service/fragments/setting-up-your-app-android.md

## Overview

> [Amazon Location Service](http://aws.amazon.com/location) is a geolocation service that lets you add location information to your applications. With Amazon Location Service, you can build applications that provide maps and points of interest, track resources, trigger actions based on location, and convert street addresses into geographic coordinates. This guide provides you with the following information:

Reviewer: For a first time reader, who is new to the domain, "assets" may elicit more of a business use case than "resources." My concern with "resources" is that it sounds like an abstract entity within a software system, as opposed to a physical object that a business owns.

```suggestion
[Amazon Location Service](http://aws.amazon.com/location) is a geolocation service that lets you add location information to your applications. With Amazon Location Service, you can build applications that provide maps and points of interest, track assets, trigger actions based on location, and convert street addresses into geographic coordinates. This guide provides you with the following information:
```

Author: I tihnk this wordsmithing is great! I'm going to hold it for now since we went over this quite a bit internally, and we can revisit later.

**To install dependencies:**

1. In Android Studio, expand **Gradle Scripts** in the project viewer and open **build.gradle (Module: YourApp).**
1. In the **dependencies** block, add the following lines:

    ```groovy
    implementation 'com.amazonaws:aws-android-sdk-mobile-client:2.20.+'
    implementation 'com.amazonaws:aws-android-sdk-location:2.20.+'
    ```

1. Click **Sync Now** in the notification bar above the file editor to update your project’s configuration. You have now successfully added the Amazon Location Service SDK dependencies to your Android app.

## Adding authentication to your App using the Amplify CLI

The Amplify Framework uses [Amazon Cognito](https://aws.amazon.com/cognito/) as its primary authentication provider. Amazon Cognito is a robust user directory service that handles user registration, authentication, account recovery, and other operations. Use the following procedure to add authentication to your app using Amazon Cognito and Amplify CLI. 

1. Open a terminal window in Android Studio by clicking Terminal.
1. Install the [Amplify CLI](https://docs.amplify.aws/cli/start/install) by running the following command:

    ```bash

>     npm install -g @aws-amplify/cli

Reviewer: As this is the Android guide, shall we use the new-and-improved procedure, instead? To wit, `npm` had been a nearly non-existent aspect of my Android development life, before working on Amplify.

```suggestion
    curl -sL https://aws-amplify.github.io/amplify-cli/install | bash && $SHELL
```

Author: This seems valid to me. Can we make this work for Windows users, too? I think there's a tabbed snippet.

Author: Addressed!

    ```

1. Initialize a new Amplify project by running the following command from the root directory of your app:

    ```bash
    amplify init
    ```

1. Create a Cognito Identity Pool. It will be used to authenticate your app users and authorize their access to Amazon Location Service. To start provisioning authentication resources in the backend, go to your project directory and run the following command:

    ```bash
    amplify add auth
    ```

1. When prompted, provide the following information:

    ```console
    ? Do you want to use the default authentication and security configuration? 
      `Default configuration`
    ? How do you want users to be able to sign in? 
      `Username`
    ? Do you want to configure advanced settings? 
      `No, I am done.`
    ```

> 1. Run the following command to push your changes to the cloud. When completed, the awsconfiguration.json file will be updated to reference your newly provisioned backend auth resources.

Reviewer: ```suggestion
1. Run the following command to push your changes to the cloud. When completed, the `awsconfiguration.json` file will be updated to reference your newly provisioned backend auth resources.
```

    ```bash
    amplify push
    ```

## Initializing the SDK

To use your new Amazon Cognito identity pool, add the following code to initialize AWSMobileClient in your Activity's onCreate method:

<amplify-block-switcher>

<amplify-block name="Java">

```java
AWSMobileClient.getInstance().initialize(getApplicationContext(), new Callback<UserStateDetails>() {
    @Override
    public void onResult(UserStateDetails userStateDetails) {
        Log.i("QuickStart", "onResult: " + userStateDetails.getUserState());
    }

    @Override

>     public void onError(Exception e) {
>         Log.e("QuickStart", "Initialization error: ", e);

Reviewer: ```suggestion
    public void onError(Exception error) {
        Log.e("QuickStart", "Initialization error: ", error);
```

    }
});
```

</amplify-block>
<amplify-block name="Kotlin">

```kotlin
AWSMobileClient.getInstance().initialize(applicationContext, object : Callback<UserStateDetails> {
    override fun onResult(userStateDetails: UserStateDetails) {
        Log.i("QuickStart", "onResult: " + userStateDetails.userState)
    }

>     override fun onError(e: Exception) {
>         Log.e("QuickStart", "Initialization error: ", e)

Reviewer: ```suggestion
    override fun onError(error: Exception) {
        Log.e("QuickStart", "Initialization error: ", error)
```

    }
})
```

</amplify-block>
</amplify-block-switcher>

This allows guest users access to your backend resources. 

You have now successfully setup and configured your app to use Amazon Location Service!

## Using the Amazon Location Service APIs

> In order to access Amazon Location Service APIs you will need to create resources. You can create resources using the [Amazon Location Service console](http://console.aws.amazon.com/location/home) or using the [AWS Command Line Interface (CLI)](https://aws.amazon.com/cli/). Once you have created the resources, you can use these resources by calling the [APIs](https://aws-amplify.github.io/aws-sdk-android/docs/reference/). 

Reviewer 2: We haven't updated our javadoc (https://aws-amplify.github.io/aws-sdk-android/docs/reference/) since 2.16.12, so it doesn't have any Location API documentation.  I think we either need to remove this link, or get the javadoc up to date (I don't know what the process is for that, yet)

Author: @richardmcclellan resolved this 👍 .

1. You will be directed to the Amazon Cognito console. Click on **Edit identity pool** in the top right corner of the page.
1. Open the drop down for **Unauthenticated identities**, choose **Enable access to unauthenticated identities**, and then press **Save Changes**.
1. Click on **Edit identity pool** once more. Make a note of the name of the Unauthenticated role. For example, `amplify-<project_name>-<env_name>-<id>-unauthRole`.
1. Open the [AWS Identity and Access Management (IAM) console](https://console.aws.amazon.com/iam/home#/roles) to manage Roles.
1. In the **Search** field, enter the name of your `unauthRole` noted above and click on it.
1. Click **+Add inline policy**, then click on the **JSON** tab.
1. Fill in the `[ARN]` placeholder with the ARN of your place index which you noted above and replace the contents of the policy with the below.

    ```json
    {
        "Version": "2012-10-17",
        "Statement": [
            {
                "Effect": "Allow",
                "Action": "geo:SearchPlaceIndexForText",
                "Resource": "[ARN]"
            }
        ]
    }
    ```

1. Click on the **Review policy** button.

> 2. In the **Name** field, enter *LocationTracker*.
> 3. Click on the **Create policy** button. You have now successfully added authentication to your Android app.

Reviewer: ```suggestion
1. In the **Name** field, enter *LocationTracker*.
1. Click on the **Create policy** button. You have now successfully added authentication to your Android app.
```

1. In the **Search** field, enter the name of your Unauthenticated role noted above and click on it.
1. Click **+Add inline policy**, then click on the **JSON** tab.
1. Fill in the `[ARN]` placeholder with the ARN of your place index which you noted above and replace the contents of the policy with the below.

    ```json
{
        "Version": "2012-10-17",
        "Statement": [
            {
                "Effect": "Allow",
                "Action": "geo:SearchPlaceIndexForText",
                "Resource": "[ARN]"
            }
        ]
}
    ```

1. Click on the **Review policy** button.
1. In the **Name** field, enter *LocationTracker*.
1. Click on the **Create policy** button. You have now successfully added authentication to your Android app.

### Searching for places

Here is how you can search for places using the place index you just created:

<amplify-block-switcher>

<amplify-block name="Java">

```java
AWSMobileClient.getInstance().initialize(getApplicationContext(), new Callback<UserStateDetails>() {
    @Override
    public void onResult(UserStateDetails userStateDetails) {

>         ExecutorService executorService = Executors.newFixedThreadPool(4);

Reviewer 3: question, why is this needed and why 4?

Reviewer: Previously, we had a construct like:
```java
new Thread(() -> {
    // do location stuff ...
}).start();
```
Android's official documentation [suggested doing background tasks on an `Executor`](https://developer.android.com/guide/background/threading#executing-in-a-background-thread), and chooses a `Executors.newFixedThreadPool(4)` somewhat at random.

I think the goal here was to show some background work in a way that was close to the platforms docs.
